staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -301,7 +300,8 @@ func RecordLongRunning(req *http.Request, requestInfo *request.RequestInfo, comp
 func MonitorRequest(req *http.Request, verb, group, version, resource, subresource, scope, component, contentType string, httpCode, respSize int, elapsed time.Duration) {
 	reportedVerb := cleanVerb(verb, req)
 	dryRun := cleanDryRun(req.URL)
-	client := cleanUserAgent(utilnet.GetHTTPClient(req))
+	// blank out client string here, in order to avoid cardinality issues
```

Comment:
I'm happy approving this but an alternative is to fix the cleanUserAgent method such that it can only return N options (e.g., Browser, kubectl, unknown).

Reply:
Is it better or worse to not add this label at all, rather than add it with a dummy (empty string) value?

---

Alternative approach: Hash it to one of an acceptable number of values (100?) - with a non-secret hash function.  You can't work out what the original client value was, but the metric would still give you _some_ breakdown of query by client.

Note if we do some sort of cleaning, another useful option is to aggressively truncate the value.  The current code (by experiment on an older cluster) preserves slightly over 1MB of characters here, which is extremely excessive.  100 characters probably preserves all the relevant useful information here.

Reply:
Yup, that'd be cool too, but from talking with Han it seems like this isn't a useful metric anyway.